src/ansi-c/c_typecheck_initializer.cpp

```diff
@@ -663,10 +663,15 @@ void c_typecheck_baset::increment_designator(designatort &designator)
       assert(components.size()==entry.size);
 
       // we skip over any padding or code
-      while(entry.index<entry.size &&
+      // we also skip over anonymous members
+      while(entry.index < entry.size &&
             (components[entry.index].get_is_padding() ||
-             components[entry.index].type().id()==ID_code))
+             (components[entry.index].get_anonymous() &&
+              components[entry.index].type().id() != ID_struct_tag) ||
```

Comment:
unions should be exempt as well:
````c
#include <assert.h>

struct S
{
  int i;
  union
  {
    int a;
    int b;
  };
  int j;
} s = { 1, { 2 }, 3 };

int main()
{
  assert(s.i == 1);
  assert(s.a == 2);
  assert(s.b == 2);
  assert(s.j == 3);
  return 0;
}
````
(and it may be worth adding such a test)

Reply:
done